llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp

```diff
@@ -386,19 +389,26 @@ Instruction *MemCpyOptPass::tryMergingIntoMemset(Instruction *StartInst,
         continue;
     }
 
-    if (!isa<StoreInst>(BI) && !isa<MemSetInst>(BI)) {
-      // If the instruction is readnone, ignore it, otherwise bail out.  We
-      // don't even allow readonly here because we don't want something like:
-      // A[1] = 2; strlen(A); A[2] = 2; -> memcpy(A, ...); strlen(A).
-      if (BI->mayWriteToMemory() || BI->mayReadFromMemory())
+    if (BI->isVolatile() || BI->isAtomic())
+      break;
+
+    // If the instruction reads something, bail out.  We
+    // don't even allow readonly here because we don't want something like:
+    // A[1] = 2; strlen(A); A[2] = 2; -> memcpy(A, ...); strlen(A).
+    if (BI->mayReadFromMemory())
+      break;
```

Comment:
Wouldn't we want to skip non-aliasing reads as well?

Reply:
I somehow thought that skipping non-aliasing reads would impact the value being read. I'll add test coverage for it.